File: static/nodejs/test/dpm_query_builder.test.ts
import { describe, expect, test } from '@jest/globals';
import { makeDpmAgentQuery } from '../src/backends/dpm_agent/dpm_agent_client';
import {
  ClientVersion,
  Query as DpmAgentQuery,
} from '../src/backends/dpm_agent/dpm_agent_pb';
import { Backend } from '../src/backends/interface';
import { DateField, Field, StringField } from '../src/field';
import { Table } from '../src/table';

class TestBackend implements Backend {
  compile(_query: Table): Promise<string> {
    return Promise.resolve('SELECT * from "foo"');
  }

  execute<Row extends object>(_query: Table): Promise<Row[]> {
    let res: Row[] = [];
    return Promise.resolve(res);
  }
}

describe('makeDpmAgentQuery', () => {
  const backend = new TestBackend();

  const id = new StringField('id');
  const name = new StringField('name');
  const price = new Field<number>('price');
  const createdOn = new DateField('created_on');
  const table = new Table({
    backend,
    packageId: 'pkg-123',
    datasetName: 'ds-456',
    datasetVersion: '0.1.0',
    source: 'test',
    name: 'testTable',
    fields: [id, name, price, createdOn],
  });

  test('returns expected Query message for a query with selections only', () => {
    const query = table.select('id', 'name').limit(10);
    const dpmQuery = makeDpmAgentQuery(query);
    const want: DpmAgentQuery.AsObject = {
      id: {
        packageid: 'pkg-123',
        sourceid: '',
      },
      clientversion: {
        client: ClientVersion.Client.NODE_JS,
        datasetversion: '0.1.0',
        codeversion: '0.1.0',
      },
      selectfrom: 'testTable',
      selectList: [
        {
          argument: {
            field: {
              fieldname: 'id',
              tablename: '',
            },
          },
          alias: '',
        },
        {
          argument: {
            field: {
              fieldname: 'name',
              tablename: '',
            },
          },
          alias: '',
        },
      ],
      limit: 10,
      // Default values.
      dryrun: false,
      tablealias: '',
      type: DpmAgentQuery.Type.DATA,
      groupbyList: [],
      orderbyList: [],
      joinsList: [],
    };

    expect(dpmQuery.toObject(false)).toEqual(want);
  });

  test('returns expected Query message for a query with selections, and filter', () => {
    const query = table
      .select('id', 'name')
      .filter(name.like('%bah%').and(createdOn.before(new Date('2023-01-01'))))
      .limit(10);
    const dpmQuery = makeDpmAgentQuery(query);
    const want: DpmAgentQuery.AsObject = {
      id: {
        packageid: 'pkg-123',
        sourceid: '',
      },
      clientversion: {
        client: ClientVersion.Client.NODE_JS,
        datasetversion: '0.1.0',
        codeversion: '0.1.0',
      },
      selectfrom: 'testTable',
      selectList: [
        {
          argument: {
            field: {
              fieldname: 'id',
              tablename: '',
            },
          },
          alias: '',
        },
        {
          argument: {
            field: {
              fieldname: 'name',
              tablename: '',
            },
          },
          alias: '',
        },
      ],
      filter: {
        op: DpmAgentQuery.BooleanExpression.BooleanOperator.AND,
        argumentsList: [
          {
            condition: {
              op: DpmAgentQuery.BooleanExpression.BooleanOperator.LIKE,
              argumentsList: [
                {
                  field: {
                    fieldname: 'name',
                    tablename: '',
                  },
                },
                {
                  literal: {
                    string: '%bah%',
                    // Unfortunately, the <message>.AsObject type does not
                    // handle oneof fields as expected; all non-object type
                    // fields are required to be defined.
                    f32: 0,
                    f64: 0,
                    i32: 0,
                    i64: 0,
                    pb_boolean: false,
                    timestamp: 0,
                    ui32: 0,
                    ui64: 0,
                  },
                },
              ],
            },
          },
          {
            condition: {
              op: DpmAgentQuery.BooleanExpression.BooleanOperator.LT,
              argumentsList: [
                {
                  field: {
                    fieldname: 'created_on',
                    tablename: '',
                  },
                },
                {
                  literal: {
                    string: '2023-01-01',
                    f32: 0,
                    f64: 0,
                    i32: 0,
                    i64: 0,
                    pb_boolean: false,
                    timestamp: 0,
                    ui32: 0,
                    ui64: 0,
                  },
                },
              ],
            },
          },
        ],
      },
      limit: 10,
      // Default values.
      dryrun: false,
      tablealias: '',
      type: DpmAgentQuery.Type.DATA,
      groupbyList: [],
      orderbyList: [],
      joinsList: [],
    };

    expect(dpmQuery.toObject(false)).toEqual(want);
  });

  test('returns expected Query message for a query with selections, filter, aggregations', () => {
    const query = table
      .select('id', 'name', price.avg().as('avgPrice'))
      .filter(name.like('%bah%').and(createdOn.before(new Date('2023-01-01'))))
      .orderBy(['avgPrice', 'DESC'], [createdOn, 'ASC']) // Note that createdOn is not in the seletion.
      .limit(10);
    const dpmQuery = makeDpmAgentQuery(query);
    const want: DpmAgentQuery.AsObject = {
      id: {
        packageid: 'pkg-123',
        sourceid: '',
      },
      clientversion: {
        client: ClientVersion.Client.NODE_JS,
        datasetversion: '0.1.0',
        codeversion: '0.1.0',
      },
      selectfrom: 'testTable',
      selectList: [
        {
          argument: {
            field: {
              fieldname: 'id',
              tablename: '',
            },
          },
          alias: '',
        },
        {
          argument: {
            field: {
              fieldname: 'name',
              tablename: '',
            },
          },
          alias: '',
        },
        {
          argument: {
            aggregate: {
              op: DpmAgentQuery.AggregateExpression.AggregateOperator.MEAN,
              argument: {
                field: {
                  fieldname: 'price',
                  tablename: '',
                },
              },
            },
          },
          alias: 'avgPrice',
        },
      ],
      filter: {
        op: DpmAgentQuery.BooleanExpression.BooleanOperator.AND,
        argumentsList: [
          {
            condition: {
              op: DpmAgentQuery.BooleanExpression.BooleanOperator.LIKE,
              argumentsList: [
                {
                  field: {
                    fieldname: 'name',
                    tablename: '',
                  },
                },
                {
                  literal: {
                    string: '%bah%',
                    // Unfortunately, the <message>.AsObject type does not
                    // handle oneof fields as expected; all non-object type
                    // fields are required to be defined.
                    f32: 0,
                    f64: 0,
                    i32: 0,
                    i64: 0,
                    pb_boolean: false,
                    timestamp: 0,
                    ui32: 0,
                    ui64: 0,
                  },
                },
              ],
            },
          },
          {
            condition: {
              op: DpmAgentQuery.BooleanExpression.BooleanOperator.LT,
              argumentsList: [
                {
                  field: {
                    fieldname: 'created_on',
                    tablename: '',
                  },
                },
                {
                  literal: {
                    string: '2023-01-01',
                    f32: 0,
                    f64: 0,
                    i32: 0,
                    i64: 0,
                    pb_boolean: false,
                    timestamp: 0,
                    ui32: 0,
                    ui64: 0,
                  },
                },
              ],
            },
          },
        ],
      },
      groupbyList: [
        {
          field: {
            fieldname: 'id',
            tablename: '',
          },
        },
        {
          field: {
            fieldname: 'name',
            tablename: '',
          },
        },
        {
          field: {
            fieldname: 'created_on',
            tablename: '',
          },
        },
      ],
      orderbyList: [
        {
          argument: {
            aggregate: {
              op: DpmAgentQuery.AggregateExpression.AggregateOperator.MEAN,
              argument: {
                field: {
                  fieldname: 'price',
                  tablename: '',
                },
              },
            },
          },
          direction: DpmAgentQuery.OrderByExpression.Direction.DESC,
        },
        {
          argument: {
            field: {
              fieldname: 'created_on',
              tablename: '',
            },
          },
          direction: DpmAgentQuery.OrderByExpression.Direction.ASC,
        },
      ],
      limit: 10,
      // Default values.
      dryrun: false,
      tablealias: '',
      type: DpmAgentQuery.Type.DATA,
      joinsList: [],
    };

    expect(dpmQuery.toObject(false)).toEqual(want);
  });
});

File: static/python/src/test/dpm_query_builder_test.py
from datetime import date
import pytest

from ..table import Table
from ..field import DateField, Field, StringField
from ..version import CODE_VERSION
from ..backends.dpm_agent.dpm_agent_client import make_dpm_agent_query
from ..backends.dpm_agent.dpm_agent_pb2 import ClientVersion, Query as DpmAgentQuery


@pytest.fixture
def id():
    return StringField("id")


@pytest.fixture
def name():
    return StringField("name")


@pytest.fixture
def price():
    return Field("price")


@pytest.fixture
def created_on():
    return DateField("created_on")


@pytest.fixture
def table(id, name, price, created_on):
    backend = {}
    return Table(
        backend=backend,
        package_id="pkg-123",
        dataset_name="ds=456",
        dataset_version="0.1.0",
        source="test",
        name="testTable",
        fields=[id, name, price, created_on],
    )


def test_returns_expected_query_message_for_query_with_selections(table):
    query = table.select("id", "name").limit(10)
    dpm_agent_query = make_dpm_agent_query(query)
    assert dpm_agent_query

    want = DpmAgentQuery(
        id={"packageId": "pkg-123"},
        clientVersion={
            "client": ClientVersion.Client.PYTHON,
            "datasetVersion": "0.1.0",
            "codeVersion": "0.1.0",
        },
        selectFrom="testTable",
        select=[
            {"argument": {"field": {"fieldName": "id"}}},
            {"argument": {"field": {"fieldName": "name"}}},
        ],
        limit=10,
    )
    assert dpm_agent_query == want


def test_returns_expected_query_message_for_query_with_selections_and_filter(
    table, name, created_on
):
    query = (
        table.select("id", "name")
        .filter(name.like("%bah%") & (created_on < date(2023, 1, 1)))
        .limit(10)
    )
    dpm_agent_query = make_dpm_agent_query(query)
    assert dpm_agent_query

    want = DpmAgentQuery(
        id={"packageId": "pkg-123"},
        clientVersion={
            "client": ClientVersion.Client.PYTHON,
            "datasetVersion": "0.1.0",
            "codeVersion": "0.1.0",
        },
        selectFrom="testTable",
        select=[
            {"argument": {"field": {"fieldName": "id"}}},
            {"argument": {"field": {"fieldName": "name"}}},
        ],
        filter={
            "op": DpmAgentQuery.BooleanExpression.AND,
            "arguments": [
                {
                    "condition": {
                        "op": DpmAgentQuery.BooleanExpression.LIKE,
                        "arguments": [
                            {"field": {"fieldName": "name"}},
                            {"literal": {"string": "%bah%"}},
                        ],
                    }
                },
                {
                    "condition": {
                        "op": DpmAgentQuery.BooleanExpression.LT,
                        "arguments": [
                            {"field": {"fieldName": "created_on"}},
                            {"literal": {"string": "2023-01-01"}},
                        ],
                    }
                },
            ],
        },
        limit=10,
    )
    assert dpm_agent_query == want


def test_returns_expected_query_message_for_query_with_selections_filter_aggs_order(
    table, name, created_on, price
):
    query = (
        table.select("id", "name", price.avg().with_alias("avg_price"))
        .filter(name.like("%bah%") & (created_on < date(2023, 1, 1)))
        .order_by(["avg_price", "DESC"], [created_on, "ASC"])
        .limit(10)
    )
    dpm_agent_query = make_dpm_agent_query(query)
    assert dpm_agent_query

    want = DpmAgentQuery(
        id={"packageId": "pkg-123"},
        clientVersion={
            "client": ClientVersion.Client.PYTHON,
            "datasetVersion": "0.1.0",
            "codeVersion": "0.1.0",
        },
        selectFrom="testTable",
        select=[
            {"argument": {"field": {"fieldName": "id"}}},
            {"argument": {"field": {"fieldName": "name"}}},
            {
                "argument": {
                    "aggregate": {
                        "op": DpmAgentQuery.AggregateExpression.MEAN,
                        "argument": {"field": {"fieldName": "price"}},
                    }
                },
                "alias": "avg_price",
            },
        ],
        filter={
            "op": DpmAgentQuery.BooleanExpression.AND,
            "arguments": [
                {
                    "condition": {
                        "op": DpmAgentQuery.BooleanExpression.LIKE,
                        "arguments": [
                            {"field": {"fieldName": "name"}},
                            {"literal": {"string": "%bah%"}},
                        ],
                    }
                },
                {
                    "condition": {
                        "op": DpmAgentQuery.BooleanExpression.LT,
                        "arguments": [
                            {"field": {"fieldName": "created_on"}},
                            {"literal": {"string": "2023-01-01"}},
                        ],
                    }
                },
            ],
        },
        groupBy=[
            {"field": {"fieldName": "id"}},
            {"field": {"fieldName": "name"}},
            {"field": {"fieldName": "created_on"}},
        ],
        orderBy=[
            {
                "argument": {
                    "aggregate": {
                        "op": DpmAgentQuery.AggregateExpression.MEAN,
                        "argument": {"field": {"fieldName": "price"}},
                    },
                },
                "direction": DpmAgentQuery.OrderByExpression.DESC,
            },
            {
                "argument": {"field": {"fieldName": "created_on"}},
                "direction": DpmAgentQuery.OrderByExpression.ASC,
            },
        ],
        limit=10,
    )
    assert dpm_agent_query == want
